linkers: don't include absolue RPATH on cross-compiling

Ansuel opened the pull request:
There is currently a reproducible problem when cross-compiling with the inclusion of external shared library RPATH entry. Meson normally includes RPATH entry to permit the usage of the tool in the build process and later removes it on the intall phase. This might be ok and permits creating reproducible build to some degree when building on host (as we can expect the shared library are always placed on a standard directory path and have a consistent RPATH)

This doesn't apply for cross-compilation scenario where the shared library might be provided from an arbritrary directory to be later packed in the final system (for example a squashfs image)

On top of this on cross-compilation on 99% of the scenario, it's not really possible to run the just built tool for build usage as it probably target a different arch.

To permit building REAL reproducible binary, add extra logic to skip the inclusion of such library path in RPATH if we detect a cross-compilation scenario and limit the inclusion of library path in RPATH only to relative path (expected to be the ones specific to the building binary/internal shared library)

Ansuel committed:
linkers: don't include absolue RPATH on cross-compiling

There is currently a reproducible problem when cross-compiling with the
inclusion of external shared library RPATH entry. Meson normally
includes RPATH entry to permit the usage of the tool in the build process
and later removes it on the intall phase. This might be ok and permits
creating reproducible build to some degree when building on host (as we
can expect the shared library are always placed on a standard directory
path and have a consistent RPATH)

This doesn't apply for cross-compilation scenario where the shared
library might be provided from an arbritrary directory to be later
packed in the final system (for example a squashfs image)

On top of this on cross-compilation on 99% of the scenario, it's not
really possible to run the just built tool for build usage as it
probably target a different arch (unless the project provide an
exe_wrapper).

To permit building REAL reproducible binary, add extra logic to skip the
inclusion of such library path in RPATH if we detect a cross-compilation
scenario and limit the inclusion of library path in RPATH only to
relative path (expected to be the ones specific to the building
binary/internal shared library) if we detect the binaries can't be run
on the host system (by the use of can_run_host_binaries() API).

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

eli-schwartz reviewed (changes requested):
> On top of this on cross-compilation on 99% of the scenario, it's not really possible to run the just built tool for build usage as it probably target a different arch.

Assuming that nobody falls into the 1% case is an absolute deal breaker. At minimum, this PR cannot even be considered unless it continues to use rpaths when cross compiling but the resulting executables are runnable.

More generally, Meson was explicitly designed to support this use case via exe_wrapper set in your cross file, and APIs such as `meson.can_run_host_binaries()`.

It is absolutely vital to have such support, if one wishes to allow users to for example run testsuites during cross builds. It also allows things such as running cross-built compiletests (`cc.run()` and suchlike).

While I would like to aid reproducibility too, it is NEVER acceptable to prevent a subset of users from being able to build at all, just for the sake of a different subset benefiting from reproducible builds.

Ansuel committed:
interpreter: move can_run_host_binaries() to environment

To permit usage of can_run_host_binaries() in other location, move it to
environment. This will be needed in linker to decide if external library
should be included in RPATH entry.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

dnicolodi commented:
There is something that I don't grok. The usual way to build binary packages is to run `DESTDIR=dist/$package-$version meson install` and archive the content of `dist/$package-$version`. Meson removes the build RPATH entries when running `meson install`.  Therefore, there should be no reproducibility problems due to build RPATH. This works just the same for cross builds, I believe. Where is the problem that this patch tries to solve coming from?

Ansuel commented:
@eli-schwartz you want me to rebase this so the ci all runs green ? (the macos failing is unrelated)

neheb commented:
Needs a rebase.
